pkg/identity/service.go

```diff
@@ -0,0 +1,186 @@
+	if err != nil {
+		switch e := err.(type) {
+		case NotAuthenticated:
+			log.Infof("Authentication failed for %s: %s", reqIdentity, e)
+			return nil, status.Error(codes.FailedPrecondition, e.Error())
+		case InvalidToken:
+			log.Debugf("Invalid token provided for %s: %s", reqIdentity, e)
+			return nil, status.Error(codes.InvalidArgument, e.Error())
+		default:
+			msg := fmt.Sprintf("Error validating token for %s: %s", reqIdentity, e)
+			log.Error(msg)
+			return nil, status.Error(codes.Internal, msg)
+		}
+	}
+
+	// Ensure the requested identity matches the token's identity.
+	if reqIdentity != tokIdentity {
+		msg := fmt.Sprintf("Requested identity did not match provided token: requested=%s; found=%s",
+			reqIdentity, tokIdentity)
+		log.Debug(msg)
+		return nil, status.Error(codes.FailedPrecondition, msg)
+	}
+
+	// Create a certificate
+	crt, err := s.i.IssueEndEntityCrt(csr)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	crts := crt.ExtractRaw()
+	if len(crts) == 0 {
+		log.Fatal("The issuer provided a certificate with key material.")
+	}
+
+	// Bundle issuer crt with certificate so the trust path to the root can be verified.
+	log.Infof("certifying %s until %s", tokIdentity, crt.Certificate.NotAfter)
+	validUntil, err := ptypes.TimestampProto(crt.Certificate.NotAfter)
+	if err != nil {
+		log.Errorf("Invalid expiry time: %s", err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	rsp := &pb.CertifyResponse{
+		LeafCertificate:          crts[0],
+		IntermediateCertificates: crts[1:],
+
+		ValidUntil: validUntil,
+	}
+	return rsp, nil
+}
+
+func checkRequest(req *pb.CertifyRequest) (string, []byte, *x509.CertificateRequest, error) {
+	reqIdentity := req.GetIdentity()
+	if reqIdentity == "" {
+		return "", nil, nil, errors.New("missing identity")
+	}
+
+	tok := req.GetToken()
+	if len(tok) == 0 {
+		return "", nil, nil, errors.New("missing token")
+	}
+
+	der := req.GetCertificateSigningRequest()
+	if len(der) == 0 {
+		return "", nil, nil,
+			errors.New("missing certificate signing request")
+	}
+	csr, err := x509.ParseCertificateRequest(der)
+	if err != nil {
+		return "", nil, nil, err
+	}
+
+	return reqIdentity, tok, csr, nil
+}
+
+func checkCSR(csr *x509.CertificateRequest, identity string) error {
+	if len(csr.DNSNames) != 1 {
+		return errors.New("CSR must have exactly one DNSName")
+	}
+	if csr.DNSNames[0] != identity {
+		return fmt.Errorf("CSR name does not match requested identity: csr=%s; req=%s", csr.DNSNames[0], identity)
+	}
+
+	if csr.Subject.CommonName != "" {
+		return errors.New("CommonName must be empty")
+	}
+	if len(csr.EmailAddresses) > 0 {
+		return errors.New("Cannot validate email addresses")
+	}
+	if len(csr.IPAddresses) > 0 {
+		return errors.New("Cannot validate IP addresses")
+	}
+	if len(csr.URIs) > 0 {
+		return errors.New("Cannot validate URIs")
+	}
+
+	return nil
+}
+
+func (e NotAuthenticated) Error() string {
+	_ = e // satisfy linter
```

Comment:
Am pretty sure this would work for the purpose of linting:

```
func (NotAuthenticated) Error() string {
	return "Authentication token could not be authenticated"
}
```